components/script/dom/shadowroot.rs

```diff
@@ -209,6 +218,40 @@ impl ShadowRoot {
             root,
         );
     }
+
+    pub(crate) fn register_slot(&self, slot: &HTMLSlotElement) {
+        debug!("Registering slot with name={:?}", slot.Name().str());
+
+        let mut slots = self.slots.borrow_mut();
+
+        let slots_with_the_same_name = slots.entry(slot.Name()).or_default();
+
+        // Insert the slot before the first element that comes after it in tree order
+        slots_with_the_same_name.insert_pre_order(slot, self.upcast::<Node>());
+    }
+
+    pub(crate) fn unregister_slot(&self, name: DOMString, slot: &HTMLSlotElement) {
+        debug!("Unregistering slot with name={:?}", name.str());
+
+        let mut slots = self.slots.borrow_mut();
+        let Entry::Occupied(mut entry) = slots.entry(name) else {
+            panic!("slot is not registered");
+        };
+        entry.get_mut().retain(|s| slot != &**s);
+    }
+
+    /// Find the first slot with the given name among this root's descendants in tree order
+    pub(crate) fn slot_for_name(&self, name: &DOMString) -> Option<DomRoot<HTMLSlotElement>> {
+        self.slots
+            .borrow()
+            .get(name)
+            .and_then(|slots| slots.first())
+            .map(|slot| slot.as_rooted())
```

Comment:
Very minor thing, but you can avoid using an anonymous function just be doing something like `and_then(Vec::first)`.

Reply:
That doesn't work in this case. The function passed to `and_then` must take a `&Vec<T>`, but [`slice::first`](https://doc.rust-lang.org/std/primitive.slice.html#method.first) is defined on slices, so we need to implicitly deref - which is only possible using an anonymous function.